Chap_API_Storage.tex

```diff
@@ -0,0 +1,225 @@
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+% Chapter: Storage support
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+\chapter{Storage Support Definitions}
+\label{chap:api_storage}
+
+Distributed and parallel computing systems are increasingly embracing storage hierarchies to meet the diverse data management needs of applications and other systems software in a cost-effective manner.
+These hierarchies provide access to a number of distinct storage layers, with each potentially composed of different storage hardware (e.g., HDD, SSD, tape, PMEM), deployed at different locations (e.g., on-node, on-switch, on-site, WAN), and designed using different storage paradigms (e.g., file-based, object-based).
+Each of these systems offers unique performance and usage characteristics that storage system users should carefully consider to ensure the most efficient use of storage resources.
+
+PMIx enables users to better understand storage hierarchies by defining attributes that formalize storage system characteristics, state, and other parameters.
+These attributes can be queried by applications, I/O libraries and middleware, and workflow systems to discover available storage resources and to inform on which resources are most suitable for different I/O workload requirements.
+
+%%%%%%%%%%%
+\section{Storage support constants}
+
+\declarestruct{pmix_storage_type_t}
+
+The \refstruct{pmix_storage_type_t} structure is an enumerated type for specifying the type of storage system being used.
+
+\begin{constantdesc}
+%
+\declareconstitemNEW{PMIX_STORAGE_TYPE_LUSTRE}
```

Comment:
I'm not a big fan of this particular enumeration. We will have to keep extending and deprecating these members as the technology background changes. Is it really critical that the value returned from a GET with `PMIX_STORAGE_TYPE` be a standardized value? (i.e., would just returning "Lustre" without specifying the corresponding constant PMIX_STORATE_TYPE_LUSTRE="Lustre" suffice?).

Some rules on capitalization, or the use of an uri form could be helpful instead?

Reply:
I understand the concern with this particular enumeration. Are there existing types in the standard that are along the lines of what you describe (i.e., an attribute with values that are not explicitly defined by the standard)? Avoiding standardization of all of the different types of storage systems is certainly appealing from the standpoint of maintaining standard definitions, but on the other hand, it could affect portability of code that relies on this attribute if PMIx implementations don't adhere to the same naming conventions, etc. We could standardize some best practices for naming as you point out, but there could still be some inconsistencies in naming that could trip up users. For instance, we settled on PMIX_STORAGE_TYPE_GPFS for GPFS file systems, but GPFS is actually called now called Spectrum Scale -- you could imagine on some systems the PMIx implementation may refer to it as "GPFS", while another could refer it to as "SpectrumScale" or something like that.

Reply:
If we did decide to avoid defining these types, maybe we could just use language that encouraged implementors to at least use the file system "type" contained in `mtab`. For example, on Theta this information looks like this:
```
theta-fs1 on /gpfs/theta-fs1 type gpfs (rw,relatime)
mira-home on /gpfs/mira-home type gpfs (rw,relatime)
172.22.10.92@o2ib:172.22.10.93@o2ib:/snx11214 on /lus/theta-fs0 type lustre (rw,flock,lazystatfs)
```
So, the PMIx implementation could promise to return the type (e.g., "gpfs" or "lustre") for any POSIX-like file systems, which would just leave the object storage systems (e.g., DAOS, S3) to generate names for, which is more palatable.

Reply:
> So, the PMIx implementation could promise to return the type (e.g., "gpfs" or "lustre") for any POSIX-like file systems, which would just leave the object storage systems (e.g., DAOS, S3) to generate names for, which is more palatable.

I like that suggestion a lot.  One thought that came to mind was that OrangeFS isn't included in the current type enumeration, and I'm sure if we sat down with enough people, we'd come up with several other distributed filesystems that aren't enumerated here.  Your suggestion above handles the long-tail of filesystems quite nicely IMO.

While I'm thinking about it, would FUSE filesystems be a type that would ever be returned?

Reply:
> Your suggestion above handles the long-tail of filesystems quite nicely IMO.

Agreed. We'd probably want to give some thought on how to communicate to users the type "names" they should expect to get back - I'm not sure if that should be in the Standard or not? We wouldn't want it to be different across PMIx implementations.

Reply:
Okay, so I think we want to change to a string type then (probably coupled with a version string, from @SteVwonder comment from another thread below).

As for how to communicate this information, would it make sense to include as one of the "Advice to ..." notes? I think you need to communicate to both users and implementers, though? I don't know if there's language from elsewhere in the standard we could borrow from, but we really want to encourage implementers to use this 'type' value provided by the kernel, so that users can have an expectation of what a valid type string would be.

Reply:
@SteVwonder according to this: https://www.kernel.org/doc/html/latest/filesystems/fuse.html

we should be able to identify particular filesystems as being `fuse` of `fuseblk`